prov/sockets/src/sock_progress.c

```diff
@@ -1177,21 +1181,30 @@ ssize_t sock_rx_claim_recv(struct sock_rx_ctx *rx_ctx, void *context,
 	return ret;
 }
 
-static int sock_pe_progress_buffered_rx(struct sock_rx_ctx *rx_ctx)
+static int sock_pe_progress_buffered_rx(struct sock_rx_ctx *rx_ctx,
+					bool shallow)
 {
 	struct dlist_entry *entry;
 	struct sock_pe_entry pe_entry;
 	struct sock_rx_entry *rx_buffered, *rx_posted;
 	size_t i, rem = 0, offset, len, used_len, dst_offset, datatype_sz;
+	size_t max_cnt = SOCK_EP_MAX_PROGRESS_CNT;
 	char *src, *dst;
 
 	if (dlist_empty(&rx_ctx->rx_entry_list) ||
 	    dlist_empty(&rx_ctx->rx_buffered_list))
 		return 0;
 
-	for (entry = rx_ctx->rx_buffered_list.next;
-	     entry != &rx_ctx->rx_buffered_list;) {
-
+	if (!shallow) {
+		rx_ctx->progress_start = &rx_ctx->rx_buffered_list;
+		max_cnt = UINT_MAX;
+	}
+	/* continue where last time left off */
+	entry = rx_ctx->progress_start;
+	if (entry == &rx_ctx->rx_buffered_list) {
+		entry = entry->next;
+	}
```

Comment:
The above code only applies to the shallow case.

Reply:
You are right. Code modified.